src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/DataSetVersionChangeService.cs

```diff
@@ -0,0 +1,258 @@
+            {
+                Filters = filterChanges?.GetValueOrDefault(ChangeType.Major),
+                FilterOptions = filterOptionChanges?.GetValueOrDefault(ChangeType.Major),
+                GeographicLevels = geographicLevelOptionChange?.GetValueOrDefault(ChangeType.Major),
+                Indicators = indicatorChanges?.GetValueOrDefault(ChangeType.Major),
+                LocationGroups = locationGroupChanges?.GetValueOrDefault(ChangeType.Major),
+                LocationOptions = locationOptionChanges?.GetValueOrDefault(ChangeType.Major),
+                TimePeriods = timePeriodChanges?.GetValueOrDefault(ChangeType.Major)
+            },
+            MinorChanges = new ChangeSetViewModel
+            {
+                Filters = filterChanges?.GetValueOrDefault(ChangeType.Minor),
+                FilterOptions = filterOptionChanges?.GetValueOrDefault(ChangeType.Minor),
+                GeographicLevels = geographicLevelOptionChange?.GetValueOrDefault(ChangeType.Minor),
+                Indicators = indicatorChanges?.GetValueOrDefault(ChangeType.Minor),
+                LocationGroups = locationGroupChanges?.GetValueOrDefault(ChangeType.Minor),
+                LocationOptions = locationOptionChanges?.GetValueOrDefault(ChangeType.Minor),
+                TimePeriods = timePeriodChanges?.GetValueOrDefault(ChangeType.Minor)
+            }
+        };
+    }
+
+    private static Dictionary<ChangeType, List<IndicatorChangeViewModel>>? GetIndicatorChanges(
+        List<IndicatorMetaChange> changes)
+    {
+        return changes.Count != 0
+            ? changes
+                .Select(c => new IndicatorChangeViewModel
+                {
+                    CurrentState = c.CurrentState is not null
+                        ? IndicatorViewModel.Create(c.CurrentState)
+                        : null,
+                    PreviousState = c.PreviousState is not null
+                        ? IndicatorViewModel.Create(c.PreviousState)
+                        : null,
+                })
+                .GroupBy(c => c.IsMajor() ? ChangeType.Major : ChangeType.Minor)
+                .ToDictionary(g => g.Key, g => g.ToList())
+            : null;
+    }
+
+    private static Dictionary<ChangeType, List<FilterChangeViewModel>>? GetFilterChanges(List<FilterMetaChange> changes)
+    {
+        return changes.Count != 0
+            ? changes
+                .Select(c => new FilterChangeViewModel
+                {
+                    CurrentState = c.CurrentState is not null
+                        ? FilterViewModel.Create(c.CurrentState)
+                        : null,
+                    PreviousState = c.PreviousState is not null
+                        ? FilterViewModel.Create(c.PreviousState)
+                        : null,
+                })
+                .GroupBy(c => c.IsMajor() ? ChangeType.Major : ChangeType.Minor)
+                .ToDictionary(g => g.Key, g => g.ToList())
+            : null;
```

Comment:
I was wondering if there is a generic method like `GetChanges` that could be abstracted out here. Did you already try anything like this?

Was thinking about something like this, but this would need some further work to make it compatible:

```csharp
    private static Dictionary<ChangeType, List<ChangeViewModel<TViewModel>>>? GetChanges<TViewModel, TEntity>(
        List<Change<TEntity>> changes,
        Func<Change<TEntity>, ChangeViewModel<TViewModel>> selector)  =>
        changes.Count != 0
            ? changes
                .Select(selector)
                .GroupBy(c => c.IsMajor() ? ChangeType.Major : ChangeType.Minor)
                .ToDictionary(g => g.Key, g => g.ToList())
            : null;
```

Reply:
Gave this a bit of a shot, but the generics are pretty gnarly as `Change` and `ChangeViewModel` themselves have their own type parameters.

Didn't really think it was worth pursuing this further as a bit of copy and paste is fine in this case.

Reply:
Ok, makes sense. Thanks for giving it a try! 🙂